server/matchmaker/algorithm/stable_marriage.py

```diff
@@ -0,0 +1,262 @@
+
+
+@with_logger
+class _MatchingGraph:
+    @staticmethod
+    def build_full(searches: List[Search]) -> WeightedGraph:
+        """A graph in adjacency list representation, whose nodes are the searches
+        and whose edges are the possible matchings for each node. Checks every
+        possible edge for inclusion in the graph.
+
+        Note that the highest quality searches come at the end of the list so that
+        it can be used as a stack with .pop().
+
+        Time complexity: O(n^2)
+        """
+        adj_list = {search: [] for search in searches}
+
+        # Generate every edge. There are 'len(searches) choose 2' of these.
+        for search, other in itertools.combinations(searches, 2):
+            quality = search.quality_with(other)
+            if not _MatchingGraph.is_possible_match(search, other, quality):
+                continue
+
+            # Add the edge in both directions
+            adj_list[search].append((other, quality))
+            adj_list[other].append((search, quality))
+
+        # Sort edges by their weights i.e. match quality
+        for search, neighbors in adj_list.items():
+            neighbors.sort(key=lambda edge: edge[1])
+
+        return adj_list
+
+    @staticmethod
+    def build_fast(searches: List[Search]) -> WeightedGraph:
+        """Builds approximately the same graph as `build_full`, but does not
+        check every possible edge.
+
+        Time complexity: O(n*log(n))
+        """
+        adj_list = {search: [] for search in searches}
+        # Sort all searches by players average trueskill mean
+        searches = sorted(searches, key=avg_mean)
+        # Now compute quality with `num_to_check` nearby searches on either side
+        num_to_check = int(math.log(max(16, len(searches)), 2)) // 2
+        for i, search in enumerate(searches):
+            for other in searches[i + 1: i + 1 + num_to_check]:
+                quality = search.quality_with(other)
+                if not _MatchingGraph.is_possible_match(search, other, quality):
+                    continue
+
+                # Add the edge in both directions
+                adj_list[search].append((other, quality))
+                adj_list[other].append((search, quality))
+
+        # Sort edges by their weights i.e. match quality
+        for search, neighbors in adj_list.items():
+            neighbors.sort(key=lambda edge: edge[1])
+
+        return adj_list
+
+    @staticmethod
+    def is_possible_match(search: Search, other: Search, quality: float) -> bool:
+        log_string = "Quality between %s and %s: %.3f thresholds: [%.3f, %.3f]."
+        log_args = (
+            search,
+            other,
+            quality,
+            search.match_threshold,
+            other.match_threshold,
+        )
+
+        if search._match_quality_acceptable(other, quality):
+            _MatchingGraph._logger.debug(
+                f"{log_string} Will be considered during stable marriage.", *log_args
```

Comment:
Not a fan of this f-string stuff with the logger. Imo the duplication is fine here.